nncf/quantization/waveq_loss.py

```diff
@@ -0,0 +1,67 @@
+import math
+import torch
+
+from nncf.compression_method_api import CompressionLoss
+from nncf.quantization.layers import QUANTIZATION_MODULES
+from nncf.utils import get_all_modules_by_type
+from nncf.nncf_network import NNCFNetwork
+
+
+class WaveQLoss(CompressionLoss):
+
+    def __init__(self, model: NNCFNetwork):
+        self.model = model
+        # upper line doesn't work
+        # "cannot assign module before Module.__init__() call")
+        self.modules = model.children()
+        self.quantize_modules = self.model_to_quantize_modules_convert(self.model)
+        self.hook_handlers = None
+
+    def model_to_quantize_modules_convert(self, model):
+        for class_type in QUANTIZATION_MODULES.registry_dict.values():
+            quantization_type = class_type.__name__
+            module_dict = get_all_modules_by_type(self.model, quantization_type)
+        return module_dict
+
+    def calc_quant_tensor_by_module(self):
+        self.hook_handlers = []
+        for module in self.quantize_modules:
+            hook_holder = LossHookHolder()
+            self.hook_handlers.append(torch.nn.Module.register_forward_hook(hook_holder.calc_hook()))
+        self.model.do_dummy_forward()
```

Comment:
no, forward should be called by user from sample. you just set up `hooks` ("observers") there.